File: practice/leetcode/python/108_sorted_array_to_bst.py
#https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/description/
#Given an integer array nums where the elements are sorted in ascending order, convert it to a
#height-balanced
#binary search tree.

class TreeNode:
    def __init__ (self, val=0, left=None, right=None):
        self.val = 0
        self.left = left
        self.right = right

class Solution:
    def sortedArrayToBst(self, nums):
        def recurse(start, end):
            if end < start:
                return None

            mid = start + (end - start)/2
            root = TreeNode(nums[mid])
            root.left = recurse(start, mid)
            root.right = recurse(mid+1, end)
            return(root)
            
        return(recurse(0, len(nums)-1))

File: practice/leetcode/cpp/108_sorted_array_to_bst.cpp
//https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/description/
//Given an integer array nums where the elements are sorted in ascending order, convert it to a
//height-balanced
//binary search tree.
using namesapce std;
#include <vector>
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(): val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x): val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right): val(x), left(left), right(right) {}
};

class Solution {
public:
	TreeNode *recurse(vector<int> &nums, int left, int right) {
		int middle = 0;	
		if (right < left) 
			return nullptr;

		middle = left + ((right - left)/2 
		TreeNode *root = new TreeNode(nums[middle]);
		root->left = recurse(nums, left, middle);
		root->right = recurse(nums, middle+1, right);
		return(root);

	}
	TreeNode *arr_to_bst(vector<int> &nums) {
		return recurse(nums,0, nums.size() - 1); 
	}
};
